### PR TITLE
Governance `drep` and `vote` command groups

### PR DESCRIPTION
**Pull request by newhoggy**
# Changelog

```yaml
- description: |
    Governance `drep` and `vote` command groups
# uncomment types applicable to the change:
  type:
  # - feature        # introduces a new feature
  - breaking       # the API has changed in a breaking way
  # - compatible     # the API has changed but is non-breaking
  # - optimisation   # measurable performance improvements
  # - improvement    # QoL changes e.g. refactoring
  # - bugfix         # fixes a defect
  # - test           # fixes/modifies tests
  # - maintenance    # not directly related to the code
  # - release        # related to a new release preparation
  # - documentation  # change in code docs, haddocks...
```

# Context

This changes the `conway governance` command structure to look like this:

```
conway governance
conway governance action
conway governance action create-constitution
conway governance action create-new-committee
conway governance action create-no-confidence
conway governance action create-protocol-parameters-update
conway governance action create-treasury-withdrawal
conway governance committee
conway governance committee create-cold-key-resignation-certificate
conway governance committee create-hot-key-authorization-certificate
conway governance committee key-gen-cold
conway governance committee key-gen-hot
conway governance committee key-hash
conway governance drep
conway governance drep delegation-certificate
conway governance drep key-gen
conway governance drep registration-certificate
conway governance vote create
```

Some types and functions in affected `conway governance` modules were moved out of those modules to facilitate code sharing of those types between those modules.

See the `cli` files for details.

# Checklist

- [ ] Commit sequence broadly makes sense and commits have useful messages
- [ ] The change log section in the PR description has been filled in
- [ ] New tests are added if needed and existing tests are updated.  These may include:
  - golden tests
  - property tests
  - round trip tests
  - integration tests
  See [Running tests](https://github.com/input-output-hk/cardano-node-wiki/wiki/Running-tests) for more details
- [ ] The version bounds in `.cabal` files are updated
- [ ] CI passes. See note on CI.  The following CI checks are required:
  - [ ] Code is linted with `hlint`.  See `.github/workflows/check-hlint.yml` to get the `hlint` version
  - [ ] Code is formatted with `stylish-haskell`.  See `.github/workflows/stylish-haskell.yml` to get the `stylish-haskell` version
  - [ ] Code builds on Linux, MacOS and Windows for `ghc-8.10.7` and `ghc-9.2.7`
- [ ] The changelog section in the PR is updated to describe the change
- [ ] Self-reviewed the diff

### CODE REVIEW

**Review by disassembler — APPROVED** [review]

File: cardano-cli/test/cardano-cli-golden/files/golden/help/allegra_governance_drep_delegation-certificate.cli [hunk 1]
```diff
@@ -0,0 +1,31 @@
+Usage: cardano-cli allegra governance drep delegation-certificate
```
Comment (carbolymer) [inline cardano-cli/test/cardano-cli-golden/files/golden/help/allegra_governance_drep_delegation-certificate.cli:1]: I was told this is wrong, so I'm not sure if `governance drep` should appear everywhere: https://input-output-rnd.slack.com/archives/G011N23CEAE/p1691759918638949?thread_ts=1691691053.122579&cid=G011N23CEAE
cc: @CarlosLopezDeLara

File: cardano-cli/src/Cardano/CLI/EraBased/Options/Governance/DRep.hs [hunk 2]
```diff
@@ -0,0 +1,179 @@
...
+                     <$> pStakePoolRegistrationParserRequirements envCli'
+                 , ShelleyToBabbageStakeKeyRegTarget sToB
+                     <$> pStakeIdentifier
+                 ]
+        <*> pOutputFile
+
+    AnyEraDeciderConwayOnwards cOn ->
+      GovernanceDRepRegistrationCertificateCmd . ConwayOnwardRegTarget cOn
+        <$> asum [ RegisterStakePool cOn
+                     <$> pStakePoolRegistrationParserRequirements envCli'
```
Comment (Jimbo4350) [inline cardano-cli/src/Cardano/CLI/EraBased/Options/Governance/DRep.hs:78]: This is not just a DRep command

File: cardano-cli/src/Cardano/CLI/EraBased/Options/Governance/DRep.hs [hunk 3]
```diff
@@ -0,0 +1,179 @@
...
+        ConwayOnwardDelegTarget cOnwards
+          <$> pStakeTarget cOnwards
+
+-- TODO: Conway era AFTER sancho net. We probably want to
+-- differentiate between delegating voting stake and reward stake
+pStakeTarget :: ConwayEraOnwards era -> Parser (StakeTarget era)
+pStakeTarget cOnwards =
+  asum
+    [ TargetStakePool cOnwards <$> pStakePoolVerificationKeyOrHashOrFile
+    , TargetVotingDrep cOnwards <$> pDRepVerificationKeyOrHashOrFile
```
Comment (Jimbo4350) [inline cardano-cli/src/Cardano/CLI/EraBased/Options/Governance/DRep.hs:142]: Same here, it's not just a DRep command

**Review by Jimbo4350 — APPROVED** [review]

File: cardano-cli/src/Cardano/CLI/EraBased/Options/Governance/DRep.hs (hunk 2 above)
Comment (newhoggy) [reply cardano-cli/src/Cardano/CLI/EraBased/Options/Governance/DRep.hs:78]: Thanks!  Will revisit this after finding more about this requirement.

File: cardano-cli/test/cardano-cli-golden/files/golden/help/allegra_governance_drep_delegation-certificate.cli (hunk 1 above)
Comment (newhoggy) [reply cardano-cli/test/cardano-cli-golden/files/golden/help/allegra_governance_drep_delegation-certificate.cli:1]: I'm following the "Cardano-CLI Feature Requirements to support Voltaire" document.

Comment (CarlosLopezDeLara) [reply cardano-cli/test/cardano-cli-golden/files/golden/help/allegra_governance_drep_delegation-certificate.cli:1]: Definitely Dreps only exist in Conway. Not in previous eras.

**Pull request merged**